File: seulho/크레인_인형뽑기.py
def solution(board, moves):
    N = len(board)
    answer = 0
    basket_stack = []
    for moveJ in moves:
        for i in range(N):
            doll = board[i][moveJ - 1]

            if doll:
                board[i][moveJ - 1] = 0
                basket_stack.append(doll)
                cnt = countDelDoll(basket_stack)
                answer += cnt
                break

    return answer


def countDelDoll(basket_stack):
    cnt = 0
    while True:
        if len(basket_stack) < 2 or basket_stack[-1] != basket_stack[-2]:
            return cnt
        basket_stack.pop()
        basket_stack.pop()
        cnt += 2

File: seulho/크레인_인형뽑기.js
function solution(board, moves) {
  N = board.length;
  let answer = 0;
  const basketStack = [];
  for (let moveJ of moves) {
    for (let i = 0; i < N; i++) {
      const doll = board[i][moveJ - 1];
      if (doll) {
        board[i][moveJ - 1] = 0;
        basketStack.push(doll);
        answer += countDelDoll(basketStack);
        break;
      }
    }
  }
  return answer;
}

function countDelDoll(basketStack) {
  let cnt = 0;
  while (true) {
    if (
      basketStack.length < 2 ||
      basketStack[basketStack.length - 1] != basketStack[basketStack.length - 2]
    ) {
      return cnt;
    }
    basketStack.pop();
    basketStack.pop();
    cnt += 2;
  }
}

const board = [
  [0, 0, 0, 0, 0],
  [0, 0, 1, 0, 3],
  [0, 2, 5, 0, 1],
  [4, 2, 4, 4, 2],
  [3, 5, 1, 3, 1],
];

const moves = [1, 5, 3, 5, 1, 2, 1, 4];

console.log(solution(board, moves));
